### ISSUE TITLE
Symbol not found: _SSL_load_error_strings

### ISSUE THREAD

**Issue by lfender6445** [libgit2/rugged#475]
running yosemite on mac - I'm using a gem bugspots (0.2.1) that depends on rugged  0.23.0b1 - i've installed cmake and pkg-config via brew to install rugged but I am getting this error:

```
dyld: lazy symbol binding failed: Symbol not found: _SSL_load_error_strings
  Referenced from: /Users/lfender_mbp/.gem/ruby/2.1.3/extensions/x86_64-darwin-14/2.1.0-static/rugged-0.23.0b1/rugged/rugged.bundle
  Expected in: flat namespace

dyld: Symbol not found: _SSL_load_error_strings
  Referenced from: /Users/lfender_mbp/.gem/ruby/2.1.3/extensions/x86_64-darwin-14/2.1.0-static/rugged-0.23.0b1/rugged/rugged.bundle
  Expected in: flat namespace

error: git-bugspots died of signal 5
```

the error itself is a bit obscure. can you point me in the right direction?

**Comment by adrianpike** [comment]
I solved a similar problem;
```
openssl.bundle, 9): Symbol not found: _SSLv2_client_method (LoadError)
  Referenced from: /Users/adrian/.rbenv/versions/2.1.2/lib/ruby/2.1.0/x86_64-darwin13.0/openssl.bundle
  Expected in: /usr/local/opt/openssl/lib/libssl.1.0.0.dylib
```

By rebuilding my Ruby version. I was using rbenv, FWIW.

**Comment by lfender6445** [comment]
interesting. perhaps I need to upgrade my ssl pkg?

**Comment by arthurschreiber** [comment]
Are you running the stock ruby version that comes with OS X, or are you running one installed from Homebrew/rbenv/rvm?

**Comment by lfender6445** [comment]
using chruby + ruby-install

**Comment by carlosmn** [comment]
The function itself has existed on every version of libssl so an upgrade wouldn't be helpful. It's more likely that it's trying to find the wrong ssl library. `otool` can help you see where it expects to load it from.

`otool -L /Users/lfender_mbp/.gem/ruby/2.1.3/extensions/x86_64-darwin-14/2.1.0-static/rugged-0.23.0b1/rugged/rugged.bundle`

It that doesn't say it's expecting to find `/usr/lib/libssl.0.9.8.dylib`, it's likely that there's something odd in your build environment making libgit2/rugged link against the wrong thing.

**Comment by lfender6445** [comment]
```
› otool -L /Users/lfender_mbp/.gem/ruby/2.1.3/extensions/x86_64-darwin-14/2.1.0-static/rugged-0.23.0b1/rugged/rugged.bundle
/Users/lfender_mbp/.gem/ruby/2.1.3/extensions/x86_64-darwin-14/2.1.0-static/rugged-0.23.0b1/rugged/rugged.bundle:
	/usr/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.5)
	/usr/lib/libiconv.2.dylib (compatibility version 7.0.0, current version 7.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1213.0.0)
	/usr/lib/libobjc.A.dylib (compatibility version 1.0.0, current version 228.0.0)
```

**Comment by carlosmn** [comment]
That is a rather odd result. The build system did recognise that openssl was available on the system, but somehow did not add the libraries to the list of dependencies. Or maybe the new release of OS X just happens to have changed enough so our assumptions don't hold. Does your pkg-config know about openssl? What does `pkg-config --libs openssl` show?

**Comment by supermarin** [comment]
I'm confirming the issue. OSX Mavericks 10.10.3, rbenv, Ruby 2.1.4 and 2.1.5, reinstalled openssl via brew.

```
[1] pry(main)> require 'rugged'
dyld: lazy symbol binding failed: Symbol not found: _SSL_load_error_strings
  Referenced from: /Users/marinusalj/.rbenv/versions/2.1.4/lib/ruby/gems/2.1.0/extensions/x86_64-darwin-14/2.1.0-static/rugged-0.23.0b1/rugged/rugged.bundle
  Expected in: flat namespace

dyld: Symbol not found: _SSL_load_error_strings
  Referenced from: /Users/marinusalj/.rbenv/versions/2.1.4/lib/ruby/gems/2.1.0/extensions/x86_64-darwin-14/2.1.0-static/rugged-0.23.0b1/rugged/rugged.bundle
  Expected in: flat namespace
```

```
$ pkg-config --libs openssl
-lssl -lcrypto -lz
```

**Comment by zouqilin** [comment]
oh, this also happed in my mac
ruby 2.1.6(compiled)
gem install rugged -v 0.23.0b1
openssl 1.0.2a

 ~/ pkg-config --libs openssl
-lssl -lcrypto -lz

**Comment by arthurschreiber** [comment]
Can you try the changes at: https://github.com/libgit2/rugged/pull/481. This fixes the issues for me.

**Comment by arthurschreiber** [comment]
@zouqilin @supermarin @lfender6445 @adrianpike Can you please try with the latest version of rugged? This issue should be fixed now, but I'd love to get that confirmed by one of you guys.

**Comment by ifigueroap** [comment]
I just updated to Yosemite today and I'm having the same issue:

```
ismael@Macintosh:~/P/H/ghcmining (master)$ ruby grabCommits.rb 
dyld: lazy symbol binding failed: Symbol not found: _SSL_load_error_strings
  Referenced from: /Library/Ruby/Gems/2.0.0/gems/rugged-0.22.2/lib/rugged/rugged.bundle
  Expected in: flat namespace

dyld: Symbol not found: _SSL_load_error_strings
  Referenced from: /Library/Ruby/Gems/2.0.0/gems/rugged-0.22.2/lib/rugged/rugged.bundle
  Expected in: flat namespace

fish: Job 1, 'ruby grabCommits.rb ' terminated by signal SIGTRAP (Trace or breakpoint trap)
ismael@Macintosh:~/P/H/ghcmining (master)$ pkg-config openssl
ismael@Macintosh:~/P/H/ghcmining (master)$ pkg-config --libs openssl
-lssl -lcrypto -lz 
```

How do I test the "latest version of rugged"? I searched the ruby gems catalog and 0.22.2 is the latest version I can install

```
ismael@Macintosh:~/P/H/ghcmining (master)$ gem list rugged --remote --all

*** REMOTE GEMS ***

em_rugged (0.3.0, 0.2.0, 0.1.5, 0.1.3, 0.1.2, 0.1.1, 0.1.0)
gitolite-rugged (1.2.2)
rugged (0.22.2, 0.21.4, 0.21.3, 0.21.2, 0.21.0, 0.19.0, 0.16.0, 0.1.2, 0.1.1, 0.1.0, 0.0.1)
rugged-mysql (0.0.2)
```

My Ruby version is:

```
ismael@Macintosh:~/P/H/ghcmining (master)$ ruby --version
ruby 2.0.0p481 (2014-05-08 revision 45883) [universal.x86_64-darwin14]
```

It's worth mentioning that the library was working correctly under the latest OSX Mavericks...

**Comment by arthurschreiber** [comment]
Try `gem install rugged --pre`.

**Comment by ifigueroap** [comment]
Thanks, I did `sudo gem install rugged --pre` and now it works as before!
The only issue is that before I was doing 

```
repo = Rugged::Repository.discover("foo")
walker = Rugged::Walker.new(repo)
walker.sorting(Rugged::SORT_TOPO)
walker.push("master")
```

and with the new version I got an error `The given reference name 'master' is not valid (Rugged::ReferenceError)`. Now I'm doing `walker.push(repo.head.target)` and seems to be working again, but I don't know if it's referring only to the master branch...

Thanks!

**Comment by arthurschreiber** [comment]
Did you try `walker.push("refs/heads/master")`?

**Comment by ifigueroap** [comment]
It will probably work! I found out that repo.head.target points to refs/head of the current branch of the open repo... Nevermind, I don't want to go offtopic on this issue, thanks a lot! :)

**Comment by zouqilin** [comment]
i tried as follow.
gem install rugged===> ok
gem list | grep rugged===>rugged (0.22.2, 0.21.4)
irb
>> require "rugged"
dyld: lazy symbol binding failed: Symbol not found: _SSL_load_error_strings
  Referenced from: /Users/zql/Software/ruby-2.1.6/lib/ruby/gems/2.1.0/extensions/x86_64-darwin-14/2.1.0/rugged-0.22.2/rugged/rugged.bundle
  Expected in: flat namespace

dyld: Symbol not found: _SSL_load_error_strings
  Referenced from: /Users/zql/Software/ruby-2.1.6/lib/ruby/gems/2.1.0/extensions/x86_64-darwin-14/2.1.0/rugged-0.22.2/rugged/rugged.bundle
  Expected in: flat namespace

[1]    16032 trace trap  irb

**Comment by kesin** [comment]
`gem install rugged --pre` works fine for me !

**Comment by matthew-white** [comment]
I just ran into this issue myself. `gem install rugged --pre` worked!

**Comment by dankohn** [comment]
If you want to install rugged 0.23.0 from a Gemfile, use the following:
```ruby
  gem 'rugged', github: 'libgit2/rugged', ref: '233da19', submodules: true
```

**Comment by arthurschreiber** [comment]
@dankohn Ugh, I see we never pushed a gem to rubygems for 0.23.0. That sucks. I'll get that resolved asap so you don't have to install from git.

**Comment by akerl** [comment]
@arthurschreiber, have you had a chance to look into releasing 0.23.0 to rubygems?

**Comment by arthurschreiber** [comment]
@akerl I just pushed the latest version, please let me know if you face any issues.

**Comment by arthurschreiber** [comment]
I'm going to close this issue as it should be fixed with the 0.23.0 release.

**Comment by syosan** [comment]
One solution was provided here

The idea is to make a static build of curl and then paste it to you /usr/bin/ or /usr/local/bin/. Briefly:

~$ xcode-select --install
Download the latest version of curl.
cd to the directory and run

~$ ./configure --disable-shared --with-darwinssl --enable-threaded-resolver

~$ make -j `sysctl -n hw.logicalcpu_max`

Copy the compiled curl, which is located in ./src/curl to you /usr/bin/

Note: In the make command in 3, the sign "`" is indeed back quote, not single quote.

 

URL：[http://unix.stackexchange.com/questions/164637/trying-to-install-curl-with-homebrew-getting-incompatible-library-version-for](http://unix.stackexchange.com/questions/164637/trying-to-install-curl-with-homebrew-getting-incompatible-library-version-for)
